src/main/java/io/fabric8/jenkins/openshiftsync/SecretWatcher.java

```diff
@@ -0,0 +1,167 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.jenkins.openshiftsync;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import hudson.triggers.SafeTimerTask;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.api.model.SecretList;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import jenkins.util.Timer;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static io.fabric8.jenkins.openshiftsync.OpenShiftUtils.getOpenShiftClient;
+import static java.net.HttpURLConnection.HTTP_GONE;
+import static java.util.logging.Level.SEVERE;
+
+/**
+ * Watches {@link Secret} objects in Kubernetes and syncs then to Credentials in Jenkins
+ */
+public class SecretWatcher implements Watcher<Secret> {
+  private static final String LABEL_JENKINS = "jenkins";
+  private static final String VALUE_SYNC = "sync";
+
+  private final Logger logger = Logger.getLogger(getClass().getName());
+  private final String[] namespaces;
+  private Watch secretWatch;
+  private ScheduledFuture relister;
+
+  public SecretWatcher(String[] namespaces) {
+    this.namespaces = namespaces;
+  }
+
+  public synchronized void start() {
+    // lets process the initial state
+    logger.info("Now handling startup secrets!!");
+    // lets do this in a background thread to avoid errors like:
+    //  Tried proxying io.fabric8.jenkins.openshiftsync.GlobalPluginConfiguration to support a circular dependency, but it is not an interface.
+    Runnable task = new SafeTimerTask() {
+      @Override
+      public void doRun() {
+        for(String namespace:namespaces) {
+          try {
+            logger.fine("listing Secrets resources");
+            final SecretList secrets = getOpenShiftClient().secrets().inNamespace(namespace).withLabel(LABEL_JENKINS, VALUE_SYNC).list();
+            onInitialSecrets(secrets);
+            logger.fine("handled Secrets resources");
+            if (secretWatch == null) {
+              secretWatch = getOpenShiftClient().secrets().inNamespace(namespace).withLabel(LABEL_JENKINS, VALUE_SYNC).withResourceVersion(secrets.getMetadata().getResourceVersion()).watch(SecretWatcher.this);
+            }
+          } catch (Exception e) {
+            logger.log(SEVERE, "Failed to load Secrets: " + e, e);
+          }
+        }
+
+      }
+    };
+    relister = Timer.get().scheduleAtFixedRate(task, 100, 10 * 1000, TimeUnit.MILLISECONDS);
+  }
+
+  public synchronized void stop() {
+    if (relister != null && !relister.isDone()) {
```

Comment:
cool ... as a heads up, https://github.com/openshift/jenkins-sync-plugin/pull/140 will be bringing in that change